crates/runtime/src/logic.rs

```diff
@@ -556,4 +559,34 @@ impl VMHostFunctions<'_> {
 
         Ok(())
     }
+
+    /// Call the contract's `send_proposal()` function through the bridge.
+    ///
+    /// The proposal actions are obtained as raw data and pushed onto a list of
+    /// proposals to be sent to the host.
+    ///
+    /// Note that multiple actions are received, and the entire batch is pushed
+    /// onto the proposal list to represent one proposal.
+    ///
+    /// # Parameters
+    ///
+    /// * `actions_ptr` - Pointer to the start of the action data in WASM
+    ///                   memory.
+    /// * `actions_len` - Length of the action data.
+    /// * `actions_ptr` - Pointer to the start of the id data in WASM memory.
+    ///                   This should always be exactly 32 bytes in length.
+    ///
+    pub fn send_proposal(
+        &mut self,
+        actions_ptr: u64,
+        actions_len: u64,
+        id_ptr: u64,
+    ) -> VMLogicResult<()> {
+        let actions_bytes: Vec<u8> = self.read_guest_memory(actions_ptr, actions_len)?;
+        let mut proposal_id = [0; 32];
+        rand::thread_rng().fill_bytes(&mut proposal_id);
+        drop(self.with_logic_mut(|logic| logic.proposals.insert(proposal_id, actions_bytes)));
```

Comment:
```suggestion
        self.with_logic_mut(|logic| logic.proposals.insert(proposal_id, actions_bytes))?;
```

also please space out this block where you can, makes it easier to read and maintain, thanks

Reply:
This is an `Option` here, not a `Result`...

Reply:
Ah, right.. okay